python/Ganga/Utility/execute.py

```diff
@@ -142,40 +202,60 @@ def execute(command,
                          stdout=subprocess.PIPE,
                          stderr=subprocess.PIPE)
 
+    # Start the background thread to catch timeout events
     timed_out = threading.Event()
-    timer = threading.Timer(timeout, __timeout_func,
-                            args=(p, timed_out))
-    timer.deamon = True
+    timer = threading.Timer(timeout, __timeout_func, args=(p, timed_out))
+    timer.daemon = True
     started_threads = []
     if timeout is not None:
         timer.start()
         started_threads.append(timer)
 
+    # This is where we store the output
     thread_output = {}
-    if not shell:
-        ti = threading.Thread(target=__reader,
-                              args=(pkl_read, pkl_write,
-                                    thread_output, 'pkl_output'))
-        ti.deamon = True
-        ti.start()
-        started_threads.append(ti)
-
-    if update_env:
-        ev = threading.Thread(target=__reader,
-                              args=(envread, envwrite,
-                                    thread_output, 'env_output'))
-        ev.deamon = True
-        ev.start()
-        started_threads.append(ev)
-
-    #logger.debug("Executing Command:\n'%s'" % str(command))
 
+    # Execute the main command of interest
+    logger.debug("Executing Command:\n'%s'" % str(command))
     stdout, stderr = p.communicate(command)
-    timer.cancel()
 
+    # Close the timeout watching thread
+    logger.debug("stdout: %s" % stdout)
+    logger.debug("stderr: %s" % stderr)
+    timer.cancel()
     for t in started_threads:
         t.join()
 
+
+    # Decode any pickled objects from disk
+    if update_env:
+        try:
+            __reader(temp_env_file.name, thread_output, 'env_output')
+        except Exception as err:
+            logger.error("Failed to Update Env after command:")# %s" % command)
+            logger.error("Error was: %s" % err)
+            logger.error("The pickle we tried to read was in: %s" % temp_env_file.name)
+            logger.error("stdout was: %s" % stdout)
+            logger.error("stderr was: %s" % stderr)
+            raise
+                                                        
+    if not shell:
+        try:
+            __reader(temp_pkl_file.name, thread_output, 'pkl_output')
+        except Exception as err:
+            logger.error("Failed to Pickle output from command: %s" % command)
+            logger.error("Error was: %s" % err)
+            logger.error("The pickle we tried to read was in: %s" % temp_pkl_file.name)
+            logger.error("stdout was: %s" % stdout)
+            logger.error("stderr was: %s" % stderr)
+            raise
+
+    # Cleanup after ourselves
+    for file_ in [temp_env_file, temp_pkl_file]:
+        if file_:
+            os.unlink(file_.name)
```

Comment:
These files don't necessarily exist. If `shell=True` and `update_env=False` are set then neither does I think.